modules/aggregation-layers/src/contour-layer/contour-layer.js

```diff
@@ -37,160 +37,158 @@ const defaultProps = {
-      countsData: weights.count.aggregationData,
-      countsBuffer: weights.count.aggregationBuffer,
-      gridSize,
-      gridOrigin,
-      cellSize,
-      boundingBox
+      dataChanged,
+      cellSizeChanged,
+      cellSize: props.cellSize,
+      needsReProjection: dataChanged || cellSizeChanged,
+      gpuAggregation
     });
   }
 
+  // Private (Contours)
+
   _generateContours() {
-    const {gridSize, gridOrigin, cellSize, thresholdData} = this.state;
-    let {countsData} = this.state;
-    if (!countsData) {
-      const {countsBuffer} = this.state;
-      countsData = countsBuffer.getData();
-      this.setState({countsData});
+    const {numCol, numRow, boundingBox, gridOffset, thresholdData} = this.state;
+    const {count} = this.state.weights;
+    let {aggregationData} = count;
+    if (!aggregationData) {
+      aggregationData = count.aggregationBuffer.getData();
+      count.aggregationData = aggregationData;
     }
 
-    const {cellWeights} = GPUGridAggregator.getCellData({countsData});
-    // const thresholds = this.props.contours.map(x => x.threshold);
+    const {cellWeights} = GPUGridAggregator.getCellData({countsData: aggregationData});
     const contourData = generateContours({
       thresholdData,
       cellWeights,
-      gridSize,
-      gridOrigin,
-      cellSize
+      gridSize: [numCol, numRow],
+      gridOrigin: [boundingBox.xMin, boundingBox.yMin],
+      cellSize: [gridOffset.xOffset, gridOffset.yOffset]
     });
 
     // contourData contains both iso-lines and iso-bands if requested.
     this.setState({contourData});
   }
 
-  _getLineLayerProps() {
-    const {colorTrigger, strokeWidthTrigger} = this.state;
-
-    return this.getSubLayerProps({
-      id: 'contour-line-layer',
-      data: this.state.contourData.contourSegments,
-      getSourcePosition: d => d.start,
-      getTargetPosition: d => d.end,
-      getColor: this._onGetSublayerColor.bind(this),
-      getWidth: this._onGetSublayerStrokeWidth.bind(this),
-      widthUnits: 'pixels',
-      updateTriggers: {
-        getColor: colorTrigger,
-        getWidth: strokeWidthTrigger
-      }
-    });
+  _updateThresholdData(props) {
+    const {contours, zOffset} = props;
+    const count = contours.length;
+    const thresholdData = new Array(count);
+    for (let i = 0; i < count; i++) {
+      const {threshold, zIndex} = contours[i];
+      thresholdData[i] = {
+        threshold,
+        zIndex: zIndex || i,
+        zOffset
+      };
+    }
+    this.setState({thresholdData});
   }
 
-  _getSolidPolygonLayerProps() {
-    const {colorTrigger} = this.state;
-
-    return this.getSubLayerProps({
-      id: 'contour-solid-polygon-layer',
-      data: this.state.contourData.contourPolygons,
-      getPolygon: d => d.vertices,
-      getFillColor: this._onGetSublayerColor.bind(this),
-      updateTriggers: {
-        getFillColor: colorTrigger
-      }
-    });
-  }
+  // Private (Sublayers)
 
   _onGetSublayerColor(element) {
```

Comment:
This is extremely expensive (creating inline functions inside an iteration).

You should be able to do something like `contours[object.contourIndex].color`.

Same thing for `_onGetSublayerStrokeWidth`